### PR TITLE
Precompile $regex filters recursively

### PR DESCRIPTION
**Pull request by plepe**
`find({ "name": { $not: { $regex: '^S' }}})`
would result in a "b.test is not a function" error, because $regex would not
have been compiled.

### CODE REVIEW

**Comment by obeliskos** [comment]
Thanks, I will need to verify the performance impact of the function call using 'npm run benchmark' later today.  Ideally any added overhead, such as function call would be made only when $regex op exists in the query.

**Comment by plepe** [comment]
I know, that performance is critical, but how to recognize whether a `$regex` operation is present, without iterating through the object? Anyway, the precompile function is called only once per execution, so impact should be minimal.

File: src/lokijs.js [hunk 1]
```diff
@@ -3328,15 +3328,28 @@
...
       }
 
-      // for regex ops, precompile
-      if (operator === '$regex') {
-        if (Array.isArray(value)) {
-          value = new RegExp(value[0], value[1]);
-        } else if (!(value instanceof RegExp)) {
-          value = new RegExp(value);
+      // precompile recursively
+      function precompile (operator, value) {
```
[the remark this answers is not in the document]
Comment (plepe) [reply src/lokijs.js]: Yes, you are right. Can't drop the function definition though, as I need to call this function recursively ...

**Comment by plepe** [comment]
I moved the precompile function out of the find() function (and renamed it to precompileQuery). Shall I merge these commits into one push force?

File: src/lokijs.js [hunk 2]
```diff
@@ -3217,6 +3217,26 @@
...
+        if (Array.isArray(value)) {
+          value = new RegExp(value[0], value[1]);
+        } else if (!(value instanceof RegExp)) {
+          value = new RegExp(value);
+        }
+      }
+
+      if (typeof value === 'object') {
+        for (var key in value) {
+          value[key] = precompileQuery(key, value[key]);
```
Comment (obeliskos) [inline src/lokijs.js]: perhaps we could constrain this recursive call to only execute if :
```javascript
if (key === '$regex' || typeof value[key] === 'object')

File: src/lokijs.js [hunk 3]
```diff
@@ -3328,14 +3348,7 @@
...
 
-      // for regex ops, precompile
-      if (operator === '$regex') {
-        if (Array.isArray(value)) {
-          value = new RegExp(value[0], value[1]);
-        } else if (!(value instanceof RegExp)) {
-          value = new RegExp(value);
-        }
-      }
+      value = precompileQuery(operator, value);
```
Comment (obeliskos) [inline src/lokijs.js]: again, if we could constrain this to only execute when :
```javascript
if (operator === '$regex' || typeof value === 'object') {
```

Comment (obeliskos) [reply src/lokijs.js]: feel free to discuss any of the above comments, they do however improve results of our single object lookups which have primitive values as well as $eq (non $regex) ops.

File: src/lokijs.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (obeliskos) [reply src/lokijs.js]: good suggestion, thanks @Kanaye

**Comment by plepe** [comment]
You are right, I included these suggestions.

**Comment by obeliskos** [comment]
Thanks again @plepe!

**Comment by plepe** [comment]
Great, thanks!

**Pull request merged**